### PR TITLE
koordlet: pod/container PSI/CPI metircs cache to tsdb

### LINKED ISSUE #1329: [proposal]refactor pod/container PSI/CPI  metircs cache to tsdb
**What is your proposal**:
see https://github.com/koordinator-sh/koordinator/issues/1241 for full description.

**Why is this needed**:
refactor pod/container PSI/CPI metircs cache to tsdb for better performance and expansibility.

**Is there a suggested solution, if so, please add it**:

### PR DESCRIPTION
**Pull request by zshmmm**
### Ⅰ. Describe what this PR does

1. refactor: koordlet: pod/container PSI/CPI metircs cache to tsdb
2. see #1241 for full description.

### Ⅱ. Does this pull request fix one issue?

fix #1329

### V. Checklist

- [x] I have written necessary docs and comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `make test`

### CODE REVIEW

File: pkg/koordlet/metriccache/metric_cache.go [hunk 1]
```diff
@@ -80,8 +82,6 @@ type MetricCache interface {
 	GetNodeCPUInfo(param *QueryParam) (*NodeCPUInfo, error)
 	GetNodeLocalStorageInfo(param *QueryParam) (*NodeLocalStorageInfo, error)
 	GetBECPUResourceMetric(param *QueryParam) BECPUResourceQueryResult
-	GetContainerInterferenceMetric(metricName InterferenceMetricName, podUID *string, containerID *string, param *QueryParam) ContainerInterferenceQueryResult
-	GetPodInterferenceMetric(metricName InterferenceMetricName, podUID *string, param *QueryParam) PodInterferenceQueryResult
```
Comment (zwzhang0107) [inline pkg/koordlet/metriccache/metric_cache.go:84]: the `InsertPodInterferenceMetrics` can also be deprecated

File: pkg/koordlet/metriccache/metric_resources.go [hunk 2]
```diff
@@ -29,4 +29,24 @@ var (
 	PodCPUThrottledMetric       = defaultMetricFactory.New(PodMetricCPUThrottled).withPropertySchema(MetricPropertyPodUID)
 	ContainerCPUUsageMetric     = defaultMetricFactory.New(ContainerMetricCPUUsage).withPropertySchema(MetricPropertyContainerID)
 	ContainerCPUThrottledMetric = defaultMetricFactory.New(ContainerMetricCPUThrottled).withPropertySchema(MetricPropertyContainerID)
+
+	// CPI
+	ContainerCPICycleMetric       = defaultMetricFactory.New(ContainerMetricCPICycle).withPropertySchema(MetricPropertyPodUID, MetricPropertyContainerID)
+	ContainerCPIInstructionMetric = defaultMetricFactory.New(ContainerMetricCPIInstruction).withPropertySchema(MetricPropertyPodUID, MetricPropertyContainerID)
+
+	// PSI
```
Comment (zwzhang0107) [inline pkg/koordlet/metriccache/metric_resources.go:37]: better to define ContainerPSI with property: containerID, podUID, resource(CPU, Memory, IO), precision(10, 60, 300), degree(full, some)

File: pkg/koordlet/metriccache/metric_types.go [hunk 3]
```diff
@@ -89,6 +111,9 @@ var MetricPropertiesFunc = struct {
 			MetricPropertyGPUDeviceUUID: uuid,
 		}
 	},
+	PodContainer: func(podUID, containerID string) map[MetricProperty]string {
```
Comment (zwzhang0107) [inline pkg/koordlet/metriccache/metric_types.go:114]: we can define PodPSI/ContainerPSI func like:
func(id, resource_type, precision, degree)

File: pkg/koordlet/metriccache/metric_types.go [hunk 4]
```diff
@@ -54,6 +54,27 @@ const (
 
 	PodMetricCPUThrottled       MetricKind = "pod_cpu_throttled"
 	ContainerMetricCPUThrottled MetricKind = "container_cpu_throttled"
+
+	// CPI
+	ContainerMetricCPICycle       MetricKind = "container_cpi_cycle"
+	ContainerMetricCPIInstruction MetricKind = "container_cpi_instruction"
+
+	// PSI
```
Comment (zwzhang0107) [inline pkg/koordlet/metriccache/metric_types.go:62]: then we only need ContainerMetricPSI and PodMetricPSI here, and some/full, cpu/mem/io, 10/60/300 are all MetricProperty

File: pkg/koordlet/metricsadvisor/collectors/performance/performance_collector_linux.go [hunk 5]
```diff
@@ -164,25 +165,34 @@ func (p *performanceCollector) getAndStartCollectorOnSingleContainer(podParentCg
...
-		klog.Errorf("insert container cpi metrics failed, err %v", err)
+
+	pciMetrics = append(pciMetrics, pciCycle, pciInstruction)
+	appender := p.metricCache.Appender()
+	if err := appender.Append(pciMetrics); err != nil {
+		klog.ErrorS(err, "Append container cycle&instruction metrics error")
+		return
 	}
+
+	if err := appender.Commit(); err != nil {
```
Comment (zwzhang0107) [inline pkg/koordlet/metricsadvisor/collectors/performance/performance_collector_linux.go:191]: how about we let profilePerfOnSingleContainer return all metric samples and do the append/commit together after the loop `for containerStatus, podMeta := range containerStatusesMap`

File: pkg/koordlet/metricsadvisor/collectors/performance/performance_collector_linux.go [hunk 6]
```diff
@@ -227,24 +238,42 @@ func (p *performanceCollector) collectSingleContainerPSI(podParentCgroupDir stri
...
-			SomeCPUAvg10:     containerPSI.CPU.Some.Avg10,
-			SomeMemAvg10:     containerPSI.Mem.Some.Avg10,
-			SomeIOAvg10:      containerPSI.IO.Some.Avg10,
-			FullCPUAvg10:     containerPSI.CPU.Full.Avg10,
-			FullMemAvg10:     containerPSI.Mem.Full.Avg10,
-			FullIOAvg10:      containerPSI.IO.Full.Avg10,
-			CPUFullSupported: containerPSI.CPU.FullSupported,
-		},
+
+	cpuSomeAvg10, err01 := metriccache.ContainerPSISomeCPUAvg10Metric.GenerateSample(
```
Comment (zwzhang0107) [inline pkg/koordlet/metricsadvisor/collectors/performance/performance_collector_linux.go:242]: we can implement as for loops here if we just define ContainerPSI as the metric.
for cpu, mem, io:
  for some, full:
    metric := generateSample
    metricList = append(metricList, metric)
return metricList

File: pkg/koordlet/metricsadvisor/collectors/performance/performance_collector_linux.go [hunk 7]
```diff
@@ -227,24 +238,42 @@ func (p *performanceCollector) collectSingleContainerPSI(podParentCgroupDir stri
...
-		klog.Errorf("insert container psi metrics failed, err %v", err)
+
+	appender := p.metricCache.Appender()
+	psiMetrics = append(psiMetrics, cpuSomeAvg10, memSomeAvg10, ioSomeAvg10, cpuFullAvg10, memFullAvg10, ioFullAvg10, cpuFullSupported)
+	if err := appender.Append(psiMetrics); err != nil {
+		klog.ErrorS(err, "Append container psi metrics error", "namespace", pod.GetNamespace(), "pod", pod.GetName(), "container", containerStatus.Name)
+		return
+	}
+
+	if err := appender.Commit(); err != nil {
```
Comment (zwzhang0107) [inline pkg/koordlet/metricsadvisor/collectors/performance/performance_collector_linux.go:272]: also, we can do the commit together

File: pkg/koordlet/metricsadvisor/collectors/performance/performance_collector_linux.go [hunk 8]
```diff
@@ -270,29 +299,49 @@ func (p *performanceCollector) collectPodPSI() {
...
-			SomeCPUAvg10:     podPSI.CPU.Some.Avg10,
-			SomeMemAvg10:     podPSI.Mem.Some.Avg10,
-			SomeIOAvg10:      podPSI.IO.Some.Avg10,
-			FullCPUAvg10:     podPSI.CPU.Full.Avg10,
-			FullMemAvg10:     podPSI.Mem.Full.Avg10,
-			FullIOAvg10:      podPSI.IO.Full.Avg10,
-			CPUFullSupported: podPSI.CPU.FullSupported,
-		},
+
+	cpuSomeAvg10, err01 := metriccache.PodPSISomeCPUAvg10Metric.GenerateSample(
```
Comment (zwzhang0107) [inline pkg/koordlet/metricsadvisor/collectors/performance/performance_collector_linux.go:310]: this part is same with containerPSI

File: pkg/koordlet/metricsadvisor/collectors/performance/performance_collector_linux.go [hunk 9]
```diff
@@ -227,72 +241,122 @@ func (p *performanceCollector) collectSingleContainerPSI(podParentCgroupDir stri
...
+		metriccache.MetricPropertiesFunc.ContainerPSI(string(pod.GetUID()), containerStatus.ContainerID, string(metriccache.PSIResourceCPU), string(metriccache.PSIPrecision10), string(metriccache.PSIDegreeFull)), collectTime, containerPSI.CPU.Full.Avg10)
+	memFullAvg10, err05 := metriccache.ContainerPSIMetric.GenerateSample(
+		metriccache.MetricPropertiesFunc.ContainerPSI(string(pod.GetUID()), containerStatus.ContainerID, string(metriccache.PSIResourceMem), string(metriccache.PSIPrecision10), string(metriccache.PSIDegreeFull)), collectTime, containerPSI.Mem.Full.Avg10)
+	ioFullAvg10, err06 := metriccache.ContainerPSIMetric.GenerateSample(
+		metriccache.MetricPropertiesFunc.ContainerPSI(string(pod.GetUID()), containerStatus.ContainerID, string(metriccache.PSIResourceIO), string(metriccache.PSIPrecision10), string(metriccache.PSIDegreeFull)), collectTime, containerPSI.IO.Full.Avg10)
+
+	cpuFullSupported, err07 := metriccache.ContainerPSICPUFullSupportedMetric.GenerateSample(
+		metriccache.MetricPropertiesFunc.PodContainer(string(pod.GetUID()), containerStatus.ContainerID), collectTime, tools.BoolToFloat64(containerPSI.CPU.FullSupported))
```
Comment (zwzhang0107) [inline pkg/koordlet/metricsadvisor/collectors/performance/performance_collector_linux.go:260]: `MetricPropertiesFunc.PodContainer()` only used here, how about just
metriccache.ContainerPSICPUFullSupportedMetric.GenerateSample(podUID, containerID, t, v)

File: pkg/koordlet/metriccache/metric_types.go [hunk 10]
```diff
@@ -100,10 +132,19 @@ var MetricPropertiesFunc = struct {
...
-		return map[MetricProperty]string{
-			MetricPropertyGPUMinor:      minor,
-			MetricPropertyGPUDeviceUUID: uuid,
-		}
+		return map[MetricProperty]string{MetricPropertyGPUMinor: minor, MetricPropertyGPUDeviceUUID: uuid}
+	},
+	PodContainer: func(podUID, containerID string) map[MetricProperty]string {
+		return map[MetricProperty]string{MetricPropertyPodUID: podUID, MetricPropertyContainerID: containerID}
+	},
+	ContainerCPI: func(podUID, containerID, cPIResource string) map[MetricProperty]string {
```
Comment (zwzhang0107) [inline pkg/koordlet/metriccache/metric_types.go:140]: typo:
cPIResource -> cpiResource
pSIXXX -> psiXXX

File: pkg/koordlet/metricsadvisor/collectors/performance/performance_collector_linux.go [hunk 11]
```diff
@@ -99,59 +100,50 @@ func (p *performanceCollector) collectContainerCPI() {
 			containerStatusesMap[containerStat] = meta
 		}
 	}
-	// get container CPI collectors for each container
-	collectors := sync.Map{}
+	metricsChan := make(chan metriccache.MetricSample, len(containerStatusesMap)*2)
+	pciMetrics := make([]metriccache.MetricSample, 0)
```
Comment (zwzhang0107) [inline pkg/koordlet/metricsadvisor/collectors/performance/performance_collector_linux.go:104]: typo
PCI -> CPI

File: pkg/koordlet/metricsadvisor/collectors/performance/performance_collector_linux.go [hunk 12]
```diff
@@ -99,59 +100,50 @@ func (p *performanceCollector) collectContainerCPI() {
...
+	// get container PCI Metric
+	go p.getContainerPCIMetrics(containerStatusesMap, metricsChan)
```
Comment (zwzhang0107) [inline pkg/koordlet/metricsadvisor/collectors/performance/performance_collector_linux.go:106]: go p.getContainerPCIMetrics() -> p.getContainerPCIMetrics()
should save metrics only if all metrics have been coillected.

File: pkg/koordlet/metricsadvisor/collectors/performance/performance_collector_linux.go [hunk 13]
```diff
@@ -99,59 +100,50 @@ func (p *performanceCollector) collectContainerCPI() {
...
+	}
+	// save container PCI metric to tsdb
+	p.saveMetric(pciMetrics)
+
+	p.started.Store(true)
+	klog.V(5).Infof("collectContainerCPI for time window %s finished at %s, container num %d",
+		timeWindow, time.Now(), len(containerStatusesMap))
+}
+
+func (p *performanceCollector) getContainerPCIMetrics(containerStatusesMap map[*corev1.ContainerStatus]*statesinformer.PodMeta, metricsChan chan<- metriccache.MetricSample) {
```
Comment (zwzhang0107) [inline pkg/koordlet/metricsadvisor/collectors/performance/performance_collector_linux.go:119]: if metricsChan is for solving the concurrency problem, maybe it is not necessary, just let getContainerPCIMetrics return with a slice is ok, and the slice should append with sync.Mutex.

File: pkg/koordlet/metricsadvisor/collectors/performance/performance_collector_linux.go [hunk 14]
```diff
@@ -99,59 +100,50 @@ func (p *performanceCollector) collectContainerCPI() {
...
-			collectorOnSingleContainer, err := p.getAndStartCollectorOnSingleContainer(parent, status, cpuNumber)
+			collectorOnSingleContainer, err := p.getAndStartCollectorOnSingleContainer(podMeta.CgroupDir, status, cpuNumber)
 			if err != nil {
 				return
 			}
-			collectors.Store(status, collectorOnSingleContainer)
-		}(containerStatus, parentPod.CgroupDir)
+			time.Sleep(p.collectTimeWindowDuration)
+			// get single container cpi metrics
+			p.profilePerfOnSingleContainer(status, collectorOnSingleContainer, podMeta.Pod, collectTime, metricsChan)
```
Comment (zwzhang0107) [inline pkg/koordlet/metricsadvisor/collectors/performance/performance_collector_linux.go:143]: then, we do not need to do the profile here and keep it as origin.
``` golang
for containerStatus, parentPod := range containerStatusesMap {
  go func(status *corev1.ContainerStatus, parent string) {
    p.getAndStartCollectorOnSingleContainer
  }
}
time.sleep
mutex sync.Mutex
for containerStatus, podMeta := range containerStatusesMap {
  go func(status *corev1.ContainerStatus, pod *corev1.Pod) {
    instruction, cycles := p.profilePerfOnSingleContainer
    mutex.lock
    append instruction and cycle metris and to slice
    mutex.unlock
  }
}
```

File: pkg/koordlet/metricsadvisor/collectors/performance/performance_collector_linux.go [hunk 15]
```diff
@@ -198,25 +194,43 @@ func (p *performanceCollector) collectContainerPSI() {
 			containerStatusesMap[containerStat] = meta
 		}
 	}
+
+	metricsChan := make(chan metriccache.MetricSample, len(containerStatusesMap)*2)
+	psiMetrics := make([]metriccache.MetricSample, 0)
+
+	go p.getContainerPSIMetrics(containerStatusesMap, metricsChan)
```
Comment (zwzhang0107) [inline pkg/koordlet/metricsadvisor/collectors/performance/performance_collector_linux.go:201]: go p.getContainerPSIMetrics -> p.getContainerPSIMetrics

File: pkg/koordlet/metricsadvisor/collectors/performance/performance_collector_linux.go [hunk 16]
```diff
@@ -198,25 +194,43 @@ func (p *performanceCollector) collectContainerPSI() {
...
 	wg.Add(len(containerStatusesMap))
 	metrics.ResetContainerPSI()
+	collectTime := time.Now()
 	for containerStatus, podMeta := range containerStatusesMap {
 		pod := podMeta.Pod
 		cgroupDir := podMeta.CgroupDir
 		go func(parentDir string, status *corev1.ContainerStatus, pod *corev1.Pod) {
 			defer wg.Done()
-			p.collectSingleContainerPSI(parentDir, status, pod)
+			p.collectSingleContainerPSI(parentDir, status, pod, collectTime, metricChan)
```
Comment (zwzhang0107) [inline pkg/koordlet/metricsadvisor/collectors/performance/performance_collector_linux.go:227]: Similarly, we can let getContainerPSIMetrics return a slice and append here, so dose the pod PSI

**Review by FillZpp — APPROVED** [review]
/lgtm

**Pull request merged**